semantic-python/src/Language/Python/ScopeGraph.hs

```diff
@@ -0,0 +1,244 @@
+{-# LANGUAGE AllowAmbiguousTypes #-}
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE DuplicateRecordFields #-}
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE FlexibleInstances #-}
+{-# LANGUAGE GeneralizedNewtypeDeriving #-}
+{-# LANGUAGE KindSignatures #-}
+{-# LANGUAGE MultiParamTypeClasses #-}
+{-# LANGUAGE NamedFieldPuns #-}
+{-# LANGUAGE PatternSynonyms #-}
+{-# LANGUAGE RankNTypes #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE StandaloneDeriving #-}
+{-# LANGUAGE TypeApplications #-}
+{-# LANGUAGE TypeOperators #-}
+{-# LANGUAGE UndecidableInstances #-}
+{-# OPTIONS_GHC -fno-warn-orphans #-}
+
+module Language.Python.ScopeGraph
+  ( scopeGraphModule
+  ) where
+
+import           Control.Algebra (Algebra (..), handleCoercible)
+import           Control.Effect.Sketch
+import           Data.Foldable
+import           Data.Monoid
+import           Data.Name
+import           Data.Typeable
+import           GHC.Generics
+import           GHC.Records
+import           GHC.TypeLits
+import           Language.Python.Patterns
+import           ScopeGraph.Convert (Result (..), complete, todo)
+import           Source.Loc
+import qualified TreeSitter.Python.AST as Py
+
+-- This orphan instance will perish once it lands in fused-effects.
+instance Algebra sig m => Algebra sig (Ap m) where
+  alg = Ap . alg . handleCoercible
+
+-- This typeclass is internal-only, though it shares the same interface
+-- as the one defined in semantic-scope-graph.
+class Typeable t => ToScopeGraph t where
+  scopeGraph ::
+    ( Has (Sketch Name) sig m
+    )
+    => t Loc
+    -> Ap m Result
```

Comment:
FWIW, you don’t actually need to specify `Ap` here; just that there’s a `Monoid` instance for `m Result`. Then you only need to mention `Ap` when you’re extracting the thing at the end: `getAp $ scopeGraph …`.